core/src/main/java/cz/o2/proxima/scheme/SchemaDescriptors.java

```diff
@@ -0,0 +1,885 @@
+      return primitiveTypeDescriptor;
+    }
+
+    public ArrayTypeDescriptor<T> getArrayTypeDescriptor() {
+      Preconditions.checkState(
+          isArrayType(), "SchemaTypeDescriptor is not array type. Type: " + getType());
+      return arrayTypeDescriptor;
+    }
+
+    public StructureTypeDescriptor<T> getStructureTypeDescriptor() {
+      Preconditions.checkState(
+          isStructureType(), "SchemaTypeDescriptor is not structure type. Type: " + getType());
+      return structureTypeDescriptor;
+    }
+
+    @SuppressWarnings("squid:S1452")
+    public EnumTypeDescriptor<?> getEnumTypeDescriptor() {
+      Preconditions.checkState(
+          isEnumType(), "SchemaTypeDescriptor is not enum type. Type: " + getType());
+      return enumTypeDescriptor;
+    }
+
+    public boolean isStructureType() {
+      return type.equals(AttributeValueType.STRUCTURE);
+    }
+
+    public boolean isArrayType() {
+      return type.equals(AttributeValueType.ARRAY);
+    }
+
+    public boolean isEnumType() {
+      return type.equals(AttributeValueType.ENUM);
+    }
+
+    public boolean isPrimitiveType() {
+      return !(isStructureType() || isArrayType() || isEnumType());
+    }
+
+    @Override
+    public SchemaTypeDescriptor<T> toTypeDescriptor() {
+      return this;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o instanceof SchemaTypeDescriptor) {
+        @SuppressWarnings("unchecked")
+        SchemaTypeDescriptor<?> other = (SchemaTypeDescriptor<?>) o;
+        if (!type.equals(other.getType())) {
+          return false;
+        }
+        switch (getType()) {
+          case STRUCTURE:
+            return structureTypeDescriptor != null
+                && structureTypeDescriptor.equals(other.getStructureTypeDescriptor());
+          case ARRAY:
+            return arrayTypeDescriptor != null
+                && arrayTypeDescriptor.equals(other.getArrayTypeDescriptor());
+          default:
+            return primitiveTypeDescriptor != null
+                && primitiveTypeDescriptor.equals(other.getPrimitiveTypeDescriptor());
+        }
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      return super.hashCode();
+    }
+  }
+
+  /**
+   * Primitive type descriptor with simple type (eq String, Long, Integer, etc).
+   *
+   * @param <T> value type
+   */
+  public static class PrimitiveTypeDescriptor<T> extends GenericTypeDescriptor<T>
+      implements WithDefaultValue<T> {
+
+    @Nullable private final Factory<T> defaultValueFactory;
+
+    public PrimitiveTypeDescriptor(
+        AttributeValueType type, @Nullable Factory<T> defaultValueFactory) {
+      super(type);
+      this.defaultValueFactory = defaultValueFactory;
+    }
+
+    @Nullable
+    public T getDefaultValue() {
+      return defaultValueFactory != null ? defaultValueFactory.apply() : null;
```

Comment:
Is it legitimate to create the descriptor with `null` default value factory? If the default value should be `null`, can that be returned by the factory itself?

Reply:
It still can be any value, that check is there just because factory is optional. Maybe signature can be
`public Optional<T> getDefaultValue()` but don't know if it will be more clear.

Reply:
Ok, now I see your point but don't know. For Proto/avro/primitives we can't have `null` default value but `JavaSerializer` probably can return default `null` but in that case will also fail code in default implementation of `hasDefaultValue`.

Reply:
Default value was removed as we discuss on slack.